pkg/reconciler/inmemorychannel/inmemorychannel.go

```diff
@@ -0,0 +1,295 @@
+	impl                    *controller.Impl
+}
+
+var deploymentGVK = appsv1.SchemeGroupVersion.WithKind("Deployment")
+var serviceGVK = corev1.SchemeGroupVersion.WithKind("Service")
+
+// Check that our Reconciler implements controller.Reconciler.
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+func FilterWithNameAndNamespace(namespace, name string) func(obj interface{}) bool {
+	return func(obj interface{}) bool {
+		if object, ok := obj.(metav1.Object); ok {
+			return name == object.GetName() &&
+				namespace == object.GetNamespace()
+		}
+		return false
+	}
+}
+
+// NewController initializes the controller and is called by the generated code.
+// Registers event handlers to enqueue events.
+func NewController(
+	opt reconciler.Options,
+	dispatcherNamespace string,
+	dispatcherDeploymentName string,
+	dispatcherServiceName string,
+	inmemorychannelinformer messaginginformers.InMemoryChannelInformer,
+	deploymentInformer appsv1informers.DeploymentInformer,
+	serviceInformer corev1informers.ServiceInformer,
+	endpointsInformer corev1informers.EndpointsInformer,
+) *controller.Impl {
+
+	r := &Reconciler{
+		Base:                     reconciler.NewBase(opt, controllerAgentName),
+		dispatcherNamespace:      dispatcherNamespace,
+		dispatcherDeploymentName: dispatcherDeploymentName,
+		dispatcherServiceName:    dispatcherServiceName,
+		inmemorychannelLister:    inmemorychannelinformer.Lister(),
+		inmemorychannelInformer:  inmemorychannelinformer.Informer(),
+		deploymentLister:         deploymentInformer.Lister(),
+		serviceLister:            serviceInformer.Lister(),
+		endpointsLister:          endpointsInformer.Lister(),
+	}
+	r.impl = controller.NewImpl(r, r.Logger, ReconcilerName, reconciler.MustNewStatsReporter(ReconcilerName, r.Logger))
+
+	r.Logger.Info("Setting up event handlers")
+	inmemorychannelinformer.Informer().AddEventHandler(reconciler.Handler(r.impl.Enqueue))
+
+	// Set up watches for dispatcher resources we care about, since any changes to these
+	// resources will affect our Channels. So, set up a watch here, that will cause
+	// a global Resync for all the channels to take stock of their health when these change.
+	deploymentInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: FilterWithNameAndNamespace(dispatcherNamespace, dispatcherDeploymentName),
+		Handler:    r,
+	})
+	serviceInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: FilterWithNameAndNamespace(dispatcherNamespace, dispatcherServiceName),
+		Handler:    r,
+	})
+	endpointsInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: FilterWithNameAndNamespace(dispatcherNamespace, dispatcherServiceName),
+		Handler:    r,
+	})
+	return r.impl
+}
+
+// These 3 functions just cause a Global Resync of the channels, because any changes here
+// should be reflected onto the channels.
+func (r *Reconciler) OnAdd(obj interface{}) {
```

Comment:
Adding a comment like "ResourceEventHandler impl" helps a lot with readability.

Reply:
+ having a compile time check to enforce Reconciler implements ResourceEventHandler helps too - could be just after "Reconciler" is defined.

Reply:
done.